enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -1348,6 +1365,105 @@ class AdjointGenerator
       addToDiffe(orig_op1, dif1, Builder2, addingType);
   }
 
+  void createBinaryOperatorDual(llvm::BinaryOperator &BO) {
+    IRBuilder<> Builder2(&BO);
+
+    Instruction *nBO = gutils->getNewFromOriginal(&BO);
+
+    assert(nBO);
+    assert(nBO->getNextNode());
+
+    if (nBO->getNextNode()) {
+      Builder2.SetInsertPoint(nBO->getNextNode());
+    }
+
+    Builder2.SetCurrentDebugLocation(
+        gutils->getNewFromOriginal(Builder2.getCurrentDebugLocation()));
+    Builder2.setFastMathFlags(getFast());
+
+    Value *orig_op0 = BO.getOperand(0);
+    Value *orig_op1 = BO.getOperand(1);
+
+    bool constantval0 = gutils->isConstantValue(orig_op0);
+    bool constantval1 = gutils->isConstantValue(orig_op1);
+
+    Value *dif0 = constantval0 ? nullptr : diffe(orig_op0, Builder2);
+    Value *dif1 = constantval1 ? nullptr : diffe(orig_op1, Builder2);
+
+    Type *addingType = BO.getType();
+
+    switch (BO.getOpcode()) {
+    case Instruction::FMul: {
+      if (constantval0) {
+        Value *idiff = Builder2.CreateFMul(orig_op0, dif0);
+        setDiffe(&BO, idiff, Builder2);
+      } else if (constantval1) {
+        Value *idiff = Builder2.CreateFMul(orig_op1, dif1);
+        setDiffe(&BO, idiff, Builder2);
+      } else {
+        Value *idiff0 =
+            Builder2.CreateFMul(dif0, gutils->getNewFromOriginal(orig_op0));
+        Value *idiff1 =
+            Builder2.CreateFMul(dif1, gutils->getNewFromOriginal(orig_op1));
+        addToDiffe(&BO, idiff0, Builder2, addingType);
```

Comment:
Why is this dual += whereas above it is dual =.

I might also restructure this code to 

if (!constantval0) {
   diffe(op0) = ...
}
if (!constantval1) {
  diffe(op1) = ..
}

there'd be less cases that way and perhaps easier to debug.